### PR TITLE
Implement PreStop hooks.

### PR DESCRIPTION
**Pull request by brendandburns**
@dchen1107 @bgrant0607 

If this looks reasonable, I'll add unit and e2e tests...

Thanks!

### CODE REVIEW

**Comment by brendandburns** [comment]
Closes https://github.com/GoogleCloudPlatform/kubernetes/issues/7091

**Comment by brendandburns** [comment]
@piosz

**Comment by bgrant0607** [comment]
What happens in the case of Kubelet failure?

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1489,6 +1498,15 @@ func (kl *Kubelet) killUnwantedPods(desiredPods map[types.UID]empty,
 	return running, utilErrors.NewAggregate(errs)
 }
 
+func findPod(pods []*api.Pod, name string) *api.Pod {
+	for ix := range pods {
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: A map would have been better. How about a TODO for using a map?

**Comment by vishh** [comment]
Needs a rebase

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (brendandburns) [inline pkg/kubelet/kubelet.go]: added a TODO, though I'm really not sure there is measurable win here, given expected # of pods / machine.

Comment (pmorie) [inline pkg/kubelet/kubelet.go]: What is the effect of a prestop hook failing?

Comment (brendandburns) [inline pkg/kubelet/kubelet.go]: It will log an error, but the containers are still destroyed.

**Comment by brendandburns** [comment]
Rebased.

@bgrant0607 I'm not sure what you mean about a kubelet failure.  Do you mean "what happens if there is a kubelet failure between a PreStop and the actual killing of the container?"

In that case, we will call PreStop multiple times.  This isn't very clearly documented.  Would you like me to do a separate PR to update the docs to reflect this?

Thanks
--brendan

**Comment by pmorie** [comment]
@brendanburns What is the effect of a pre-stop hook failure?

**Comment by vmarmol** [comment]
Today we call killPod() before we start a new pod we've never started before. This would mean that before your pod starts we run its pre-stop hook. That's probably not what the user intended. We'll need to differentiate the first start of the pod vs. subsequent starts in order to avoid that.

**Comment by brendandburns** [comment]
@vmarmol Ah, yes, good point, I'll add that in.

**Comment by bgrant0607** [comment]
Yes, failure after delivering prestop. Yes, please update the doc.

If the containers exit after receiving prestop, will Kubelet still restart them? How about after a Kubelet failure?

@vishh What would happen to an exec-based hook running when Kubelet died?

**Comment by vmarmol** [comment]
@bgrant0607 if the container exits during the pre-stop it will be treated as any container termination. It will be restarted if the policies allow for it.

**Comment by brendandburns** [comment]
The exec hook is running inside of Docker, so it's already abstracted across an HTTP API, I think it will just complete on the Docker side. 

@vmarmol I added a check to validate that the container is actually running before we send the hook.

**Comment by brendandburns** [comment]
Added docs clarification here:
https://github.com/GoogleCloudPlatform/kubernetes/pull/7450

**Comment by bgrant0607** [comment]
@vmarmol Actually, I think the pod won't be among the desired pods, so its containers won't be restarted. We'll need to change this a bit in order to handle graceful termination when we add that.

**Comment by vmarmol** [comment]
@bgrant0607 in the case that we're deleting the pod from the node I believe that is true (but that's the desired behavior no?). The pod will be in the desired pod if we saw an update that required us to change the infra container (and thus kill and re-start the pod).

**Comment by vishh** [comment]
As Brenden mentioned, kubelet death should not an issue wrt. to docker exec. The exec'ed process will continue to run inside the container.

**Comment by pmorie** [comment]
To clarify my earlier comment, does anything happen if a prestop hook exits with !zero?

**Comment by brendandburns** [comment]
@pmorie no, other than an error being logged, no, nothing happens.  I think anything further is beyond the scope of 1.0, but is reasonable to discuss post 1.0

**Comment by vishh** [comment]
LGTM!

**Comment by bgrant0607** [comment]
Tests fail.

**Comment by brendandburns** [comment]
Rebased and tests fixed.  ptal.

Thanks!
--brendan

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -905,9 +905,34 @@ func (kl *Kubelet) pullImage(pod *api.Pod, container *api.Container) error {
...
+		}
+		if err := kl.RunStopHooks(apiPod, container); err != nil {
+			errList = append(errList, err)
+		}
+	}
+	if err := kl.containerManager.KillPod(pod); err != nil {
+		errList = append(errList, err)
+	}
+
+	if len(errList) > 0 {
```
Comment (vmarmol) [inline pkg/kubelet/kubelet.go]: nit: `return utilErrors.NewAggregate(errList)` should be sufficient

**Comment by vmarmol** [comment]
Sorry @brendandburns you'll need to rebase again. This part of the code is going through heavy refactoring right now. If you don't mind waiting, the refactoring is near its end and should be good by Friday :)

**Comment by vishh** [comment]
@vmarmol: Is it safe to rebase now?

**Comment by vmarmol** [comment]
@vishh @brendandburns yep, all PRs are in. Next changes should be minimally disruptive.

**Comment by brendandburns** [comment]
Ok, when tested in a real cluster the initial approach didn't work.  This is a significant re-implementation of a different approach.

Please take another look.  If you would rather have it as a different PR, let me know.

Thanks!
--brendan

**Comment by vmarmol** [comment]
Do we need to serialize the pod into a label? Seems like we should be able to provide the pod. Or was there an issue with that approach?

**Comment by brendandburns** [comment]
@vmarmol after a pod is requested for delete, the kublet no longer has the pod object.  To do otherwise would require checkpointing the state of pods to a file which I'd rather not do.

File: pkg/kubelet/dockertools/manager.go [hunk 3]
```diff
@@ -52,6 +53,9 @@ const (
 	podOomScoreAdj = -100
 
 	maxReasonCacheEntries = 200
+
+	kubernetesPodLabel       = "io.kubernetes.pod"
+	kubernetesContainerLabel = "io.kubernetes.container"
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/manager.go]: We already encode pod name as "io.kubernetes.pod.name" can we use "io.kubernetes.container.name" here?

WDYT of "io.kubernetes.pod.data" or something along those lines for the above?

File: pkg/kubelet/dockertools/manager.go [hunk 4]
```diff
@@ -507,6 +511,17 @@ func (dm *DockerManager) runContainer(pod *api.Pod, container *api.Container, op
 	labels := map[string]string{
 		"io.kubernetes.pod.name": namespacedName.String(),
 	}
+	if container.Lifecycle != nil && container.Lifecycle.PreStop != nil {
+		glog.V(1).Infof("Setting preStop hook")
+		// TODO: This is kind of hacky, we should really just encode the bits we need.
+		data, err := latest.Codec.Encode(pod)
+		if err != nil {
+			glog.Errorf("Failed to encode prestop hook")
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/manager.go]: nit: can we add the pod/container name?

File: pkg/kubelet/dockertools/manager.go [hunk 5]
```diff
@@ -1074,9 +1089,41 @@ func (dm *DockerManager) KillContainer(containerID types.UID) error {
...
+	var found bool
+	var preStop string
+	if inspect != nil && inspect.Config != nil && inspect.Config.Labels != nil {
+		preStop, found = inspect.Config.Labels[kubernetesPodLabel]
+	}
+	if found {
+		var pod api.Pod
+		err := latest.Codec.DecodeInto([]byte(preStop), &pod)
+		if err != nil {
+			glog.Errorf("Failed to decode prestop: %s", preStop)
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/manager.go]: nit: add pod/container name

**Comment by vmarmol** [comment]
Ah, I see now. Thanks for the explanantion @brendandburns! Looks good overall, mostly nits.

**Comment by brendandburns** [comment]
@vmarmol rebased and comments addressed.

ptal.
Thanks!
--brendan

**Comment by vmarmol** [comment]
LGTM, thanks @brendandburns! Will merge on green.

**Comment by vmarmol** [comment]
@brendandburns looks like a test failure:

```
--- FAIL: TestKillContainerInPodWithPreStop (0.03s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x10 pc=0x4ba3ac]

goroutine 30 [running]:
testing.func·006()
	/usr/local/go/src/testing/testing.go:441 +0x249
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools.(*DockerManager).RunInContainer(0xc208064780, 0xc68ea0, 0x4, 0xc208129a40, 0x2, 0x4, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools/_test/_obj_test/manager.go:1093 +0x98c
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/lifecycle.(*HandlerRunner).Run(0xc20809bf50, 0xc68ea0, 0x4, 0xc208180000, 0xc20807a800, 0xc208195ae0, 0x0, 0x0)
	/root/workspace/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/lifecycle/handlers.go:53 +0x11a
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools.(*DockerManager).killContainer(0xc208064780, 0xc68ea0, 0x4, 0x0, 0x0)
	github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools/_test/_obj_test/manager.go:1390 +0x1109
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools.(*DockerManager).KillContainerInPod(0xc208064780, 0xc766c0, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools/_test/_obj_test/manager.go:1338 +0x48c
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools.TestKillContainerInPodWithPreStop(0xc2080899e0)
	/root/workspace/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools/manager_test.go:354 +0xd89
testing.tRunner(0xc2080899e0, 0x12284c0)
	/usr/local/go/src/testing/testing.go:447 +0x134
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:555 +0xd4e

goroutine 1 [chan receive]:
testing.RunTests(0xe0c780, 0x12282e0, 0x18, 0x18, 0xc20807c501)
	/usr/local/go/src/testing/testing.go:556 +0xdc3
testing.(*M).Run(0xc20800bd60, 0x1bd05c0)
	/usr/local/go/src/testing/testing.go:485 +0xe1
main.main()
	github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools/_test/_testmain.go:154 +0x382

goroutine 5 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x1bc8020)
	/root/workspace/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x99
created by github.com/golang/glog.init·1
	/root/workspace/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x2ca

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools	0.196s
```

**Comment by brendandburns** [comment]
Weird, tests pass on my machine...  I'll try to figure out what is different.

**Comment by brendandburns** [comment]
@vmarmol fixed tests.  ptal.

Thanks
--brendan

**Pull request merged**